### PR TITLE
REGRESSION(267279@main): [ macOS wk2 ] fast/canvas/webgl/texImage2D-mse-flipY-false.html is a consistent text failure.

### PR DESCRIPTION
**Pull request by jyavenard**
#### 324ba3da764bc90bb9a08b8b763b785771fd37fb
<pre>
REGRESSION(267279@main): [ macOS wk2 ] fast/canvas/webgl/texImage2D-mse-flipY-false.html is a consistent text failure.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260867">https://bugs.webkit.org/show_bug.cgi?id=260867</a>
rdar://114642707

Reviewed by Youenn Fablet and Jer Noble.

The fix for bug 260185 (commit 267279@main) incorrectly changed the ordering
of operations (renqueue frames was called prior seeking with AVSampleBufferRenderSynchronizer).
This resulted in the incorrect video frame being painted.

In this change we restore the prior seeking workflow. This requires to
split the MediaSource::seekToTarget into two, the first one being to determine
if data is available at the seekpoint and calculating the nearest keyframe
(if using fastSeek) and then the actual SourceBuffer&apos;s loading of data
at the seek time.

Fly-By: commit 267279@main also re-introduced bug 188926 by dropping one
of the work-around. This was caught by media/media-source/media-source-seek-twice.html

Adding test to check that the right frame is displayed when the seeked
event is fired.

* LayoutTests/media/media-source/content/test-fragmented-video.png: Added.
* LayoutTests/media/media-source/media-managedmse-seek-expected.html: Added.
* LayoutTests/media/media-source/media-managedmse-seek.html: Added.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::waitForTarget):
(WebCore::MediaSource::completeSeek):
(WebCore::MediaSource::seekToTime):
(WebCore::MediaSource::seekToTarget): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::computeSeekTime):
(WebCore::SourceBuffer::seekToTime):
(WebCore::SourceBuffer::seekToTarget): Deleted.
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::computeSeekTime):
(WebCore::SourceBufferPrivate::seekToTarget): Deleted.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
(WebCore::SourceBufferPrivate::notifyClientWhenReadyForMoreSamples):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::MediaPlayerPrivateMediaSourceAVFObjC):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::seekInternal):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::maybeSeekCompleted):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setHasAvailableVideoFrame):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::seekCompleted): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::waitForTarget):
(WebCore::MediaSourcePrivateAVFObjC::seekToTime):
(WebCore::MediaSourcePrivateAVFObjC::seekToTarget): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::seekToTime):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::doSeek):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::waitForTarget):
(WebCore::MediaSourcePrivateGStreamer::seekToTime):
(WebCore::MediaSourcePrivateGStreamer::seekToTarget): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.h:
* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::seekToTarget):
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.cpp:
(WebCore::MockMediaSourcePrivate::waitForTarget):
(WebCore::MockMediaSourcePrivate::seekToTime):
(WebCore::MockMediaSourcePrivate::seekToTarget): Deleted.
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::waitForTarget):
(WebKit::RemoteMediaSourceProxy::seekToTime):
(WebKit::RemoteMediaSourceProxy::seekToTarget): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::computeSeekTime):
(WebKit::RemoteSourceBufferProxy::seekToTime):
(WebKit::RemoteSourceBufferProxy::seekToTarget): Deleted.
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::waitForTarget):
(WebKit::MediaSourcePrivateRemote::seekToTime):
(WebKit::MediaSourcePrivateRemote::seekToTarget): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::computeSeekTime):
(WebKit::SourceBufferPrivateRemote::seekToTime):
(WebKit::SourceBufferPrivateRemote::seekToTarget): Deleted.
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/267814@main">https://commits.webkit.org/267814@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/bbab1dcdd76d39ac9c14057ac55998557c518d6c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17766 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18096 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18634 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19590 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16619 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21387 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18247 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18664 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17979 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18265 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15434 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20456 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15497 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16187 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22746 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16509 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16355 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20608 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16927 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16039 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16026 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4236 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16772 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm [hunk 1]
```diff
@@ -544,36 +544,40 @@ void getSupportedTypes(HashSet<String>& types) const final
...
+void MediaPlayerPrivateMediaSourceAVFObjC::maybeSeekCompleted()
 {
     if (m_seekState == SeekCompleted)
         return;
     if (hasVideo() && !m_hasAvailableVideoFrame) {
         ALWAYS_LOG(LOGIDENTIFIER, "waiting for video frame");
         m_seekState = WaitingForAvailableFame;
         return;
-    }
+    } else
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm]: No else needed given there is a return above.

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm [hunk 2]
```diff
@@ -544,36 +544,40 @@ void getSupportedTypes(HashSet<String>& types) const final
...
-    m_seekState = SeekCompleted;
-    if (shouldBePlaying())
-        [m_synchronizer setRate:m_rate];
     if (m_synchronizerSeeking) {
         ALWAYS_LOG(LOGIDENTIFIER, "Synchronizer still seeking, bailing out");
         return;
     }
+    m_seekState = SeekCompleted;
+    if (shouldBePlaying())
+        [m_synchronizer setRate:m_rate];
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:584]: Are we reverting behaviour from the previous patch?

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm [hunk 3]
```diff
@@ -544,36 +544,40 @@ void getSupportedTypes(HashSet<String>& types) const final
...
+        m_mediaSourcePrivate->seekToTime(seekedTime, [this, weakThis] {
+            if (!weakThis)
+                return;
+            maybeSeekCompleted();
+        });
     });
 }
 
-void MediaPlayerPrivateMediaSourceAVFObjC::seekCompleted()
+void MediaPlayerPrivateMediaSourceAVFObjC::maybeSeekCompleted()
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm]: Would completeSeek be a more accurate name?

File: Source/WebCore/Modules/mediasource/MediaSource.cpp [hunk 4]
```diff
@@ -186,7 +186,7 @@ const PlatformTimeRanges& MediaSource::buffered() const
     return m_buffered;
 }
 
-void MediaSource::seekToTarget(const SeekTarget& target, CompletionHandler<void(const MediaTime&)>&& completionHandler)
+void MediaSource::waitForTarget(const SeekTarget& target, CompletionHandler<void(const MediaTime&)>&& completionHandler)
```
Comment (youennf) [inline Source/WebCore/Modules/mediasource/MediaSource.cpp:189]: Why waitForTarget and not computeSeekTime instead?

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm [hunk 5]
```diff
@@ -544,36 +544,40 @@ void getSupportedTypes(HashSet<String>& types) const final
...
 
         [m_synchronizer setRate:0 time:PAL::toCMTime(seekedTime)];
 
-        // In cases where the destination seek time precisely matches the synchronizer's existing time
-        // no time jumped notification will be issued. In this case, just notify the MediaPlayer that
-        // the seek completed successfully.
-        if (!m_synchronizerSeeking)
-            seekCompleted();
+        m_mediaSourcePrivate->seekToTime(seekedTime, [this, weakThis] {
+            if (!weakThis)
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:560]: I wonder whether we should not build a protectedThis here, given our current smart pointer policy approach. Ditto in other part of the patch where we have a weakThis in a lambda.

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm [hunk 6]
```diff
@@ -227,14 +227,22 @@ static bool MediaSourcePrivateAVFObjCHasAudio(SourceBufferPrivateAVFObjC* source
...
     else
         completionHandler(MediaTime::invalidTime());
 }
 
+void MediaSourcePrivateAVFObjC::seekToTime(const MediaTime& time, CompletionHandler<void()>&& completionHandler)
+{
+    if (m_client)
+        m_client->seekToTime(time, WTFMove(completionHandler));
+    else
+        completionHandler();
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:242]: We tend to do the rare case first (if (!m_client))

File: Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp [hunk 7]
```diff
@@ -355,15 +355,24 @@ void SourceBufferPrivateRemote::setAppendWindowEnd(const MediaTime& appendWindow
...
         return;
     }
 
-    gpuProcessConnection->connection().sendWithAsyncReply(Messages::RemoteSourceBufferProxy::SeekToTarget(target), WTFMove(completionHandler), m_remoteSourceBufferIdentifier);
+    gpuProcessConnection->connection().sendWithAsyncReply(Messages::RemoteSourceBufferProxy::ComputeSeekTime(target), WTFMove(completionHandler), m_remoteSourceBufferIdentifier);
+}
+
+void SourceBufferPrivateRemote::seekToTime(const MediaTime& time)
+{
+    auto gpuProcessConnection = m_gpuProcessConnection.get();
```
Comment (youennf) [inline Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:371]: We could get it after the isGPURunning check

**Review by youennf — APPROVED** [review]
LGTM once bots are happy.

File: Source/WebCore/Modules/mediasource/MediaSource.cpp (hunk 4 above)
Comment (jyavenard) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp:189]: The completionHandler will not be called until data has been appended to the source buffers.

There are two steps in this process:
1- Wait until there's data buffered at the SeekTarget
2- Determine what the actual seek time will be (Which you can only do once data is buffered) this is needed for fastSeek support.

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm (hunk 2 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:584]: no. But I figured there was no point resuming playback when seeking hasn't even complete yet.
Makes for easier code to read.

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm (hunk 5 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:560]: Those aren't refcounted object.

File: Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp (hunk 7 above)
Comment (jyavenard) [reply Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:371]: we could, but there are heaps of methods doing the same. would make a too big diff

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm (hunk 3 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm]: I've renamed it maybeCompleteSeek

**Review by jernoble — APPROVED** [review]
LGTM

**Pull request merged**